controllers/clusterwidenetworkpolicy_controller.go

```diff
@@ -19,54 +17,178 @@ package controllers
 func (r *ClusterwideNetworkPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	var clusterNP firewallv1.ClusterwideNetworkPolicy
-	if err := r.Get(ctx, req.NamespacedName, &clusterNP); err != nil {
-		return ctrl.Result{}, client.IgnoreNotFound(err)
+	log := r.log.WithValues("cwnp", req.Name)
+
+	var cwnps firewallv1.ClusterwideNetworkPolicyList
+	if err := r.List(ctx, &cwnps, client.InNamespace(firewallv1.ClusterwideNetworkPolicyNamespace)); err != nil {
+		return done, err
 	}
 
-	// if network policy does not belong to the namespace where clusterwide network policies are stored:
-	// update status with error message
-	if req.Namespace != firewallv1.ClusterwideNetworkPolicyNamespace {
-		r.recorder.Event(&clusterNP, corev1.EventTypeWarning, "Unapplicable", fmt.Sprintf("cluster wide network policies must be defined in namespace %s otherwise they won't take effect", firewallv1.ClusterwideNetworkPolicyNamespace))
-		return ctrl.Result{}, nil
+	return r.reconcileRules(ctx, log, cwnps)
+}
+
+func (r *ClusterwideNetworkPolicyReconciler) reconcileRules(ctx context.Context, log logr.Logger, cwnps firewallv1.ClusterwideNetworkPolicyList) (ctrl.Result, error) {
+	var f firewallv1.Firewall
+	nn := types.NamespacedName{
+		Name:      firewallName,
+		Namespace: firewallv1.ClusterwideNetworkPolicyNamespace,
+	}
+	if err := r.Get(ctx, nn, &f); err != nil {
+		return done, client.IgnoreNotFound(err)
 	}
 
-	err := clusterNP.Spec.Validate()
+	// Set CWNP requeue interval
+	if interval, err := time.ParseDuration(f.Spec.Interval); err == nil {
+		r.interval = interval
+	} else {
+		return done, fmt.Errorf("failed to parse Interval field: %w", err)
+	}
+
+	var services corev1.ServiceList
+	if err := r.List(ctx, &services); err != nil {
+		return done, err
+	}
+	nftablesFirewall := r.createFirewall(f, &cwnps, &services, r.dnsProxy, log)
+	if err := r.manageDNSProxy(f, cwnps, nftablesFirewall); err != nil {
+		return done, err
+	}
+	updated, err := nftablesFirewall.Reconcile()
 	if err != nil {
-		r.recorder.Event(&clusterNP, corev1.EventTypeWarning, "Unapplicable", fmt.Sprintf("cluster wide network policy is not valid: %v", err))
-		return ctrl.Result{}, nil
+		return done, err
 	}
 
-	return ctrl.Result{}, nil
+	if updated {
+		for _, i := range cwnps.Items {
+			o := i
+			if err := r.Update(ctx, &o); err != nil {
+				return done, err
+			}
+		}
+	}
+
+	return done, nil
+}
+
+// manageDNSProxy start DNS proxy if toFQDN rules are present
+// if rules were deleted it will stop running DNS proxy
+func (r *ClusterwideNetworkPolicyReconciler) manageDNSProxy(
+	f firewallv1.Firewall, cwnps firewallv1.ClusterwideNetworkPolicyList, nftablesFirewall FirewallInterface,
+) (err error) {
+	// Skipping is needed for testing
+	if r.skipDNS {
+		return nil
+	}
+
+	enableDNS := len(cwnps.GetFQDNs()) > 0
+
+	nftablesFirewall.ReconcileNetconfTables()
```

Comment:
Are there any reasons why not to run this line only when `enableDNS` is true? This line has the potential to change existing firewall configuration and break something, independent if this feature is enabled or not. I think it makes it safer to merge this, if we could move it there. Then we have a guarantee that old firewalls do not break for sure.

Probably we could also remove the `skipDNS` flag then.

Reply:
Reconcilation should be done in both cases. Firewall gets CWNPs as an argument and determines by their presence what rules should be generated. Therefore, it should be run both when CWNPs are present to create set rules and when their not present to clear old set rules.